state/table.h

```diff
@@ -0,0 +1,130 @@
+
+typedef struct table_callback_data table_callback_data;
+
+typedef void * table_done_cb;
+
+/* The callback called when the database operation hasn't completed after
+ * the number of retries specified for the operation. */
+typedef void (*table_fail_cb)(unique_id id, void *user_context);
+
+typedef void (*table_retry_cb)(table_callback_data *cb_data);
+
+struct table_callback_data {
+  /** ID of the entry in the table that we are going to look up, remove or add. */
+  unique_id id;
+  /** The callback that will be called when results is returned. */
+  table_done_cb done_cb;
+  /** The callback that will be called when the redis command times out. */
+  table_fail_cb fail_cb;
+  /** The callback that will be called to initiate the next try. */
+  table_retry_cb retry_cb;
+  /** Number of retries left. */
+  int retry_count;
+  /** Timeout, in milliseconds. */
+  uint64_t timeout;
+  /** Pointer to the data that is entered into the table. */
+  void *data;
+  /** Pointer to the data used internally to handle multiple database requests. */
+  void *requests_info;
+  /** User context. */
+  void *user_context;
+  /** Handle to db. */
+  db_handle *db_handle;
+  /** Handle to timer. */
+  int64_t timer_id;
+};
+
+/**
+ * Function to handle the timeout event.
+ * @param loop Event loop.
+ * @param timer_id Timer identifier.
+ * @param context Pointer to the callback data for the object table
+ * @return Timeout to reset the timer if we need to try again, or
+ *         EVENT_LOOP_TIMER_DONE if retry_count == 0.
+ */
+int64_t table_timeout_handler(event_loop *loop,
+                              int64_t timer_id,
+                              void *context);
+
+/**
+ *
+ * @param db_handle Database handle.
+ * @param id ID of the object that is looked up, added or removed.
+ * @param retry_count Number of retries to the database before giving up.
+ * @param timeout Timeout between retries (in milliseconds).
+ * @param data Data entered into the table.
+ * @param done_cb Function to be called when database returns result.
+ * @param fail_cb Function to be called when number of retries is exhausted.
+ * @param retry_cb Function that will be called to initiate the next try.
+ * @param user_context Context that can be provided by the user and will be
+ *        passed on to the various callbacks.
+ * @return New table callback data struct.
+ */
+table_callback_data *init_table_callback(db_handle *db_handle,
+                                         unique_id id,
+                                         int retry_count,
+                                         uint64_t timeout,
+                                         void *data,
+                                         table_done_cb done_cb,
+                                         table_fail_cb fail_cb,
+                                         table_retry_cb retry_cb,
+                                         void *user_context);
+
+
+void destroy_table_callback(table_callback_data *cb_data);
```

Comment:
Documentation? And the difference between this and `outstanding_callbacks_remove`.